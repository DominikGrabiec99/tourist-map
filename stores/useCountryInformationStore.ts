import { defineStore } from 'pinia';

/** INTERFACES */
import ICountryInformation from '@/ts/interfaces/CountryInformation';
import IExchangeRates from '@/ts/interfaces/ExchangeRates';

interface IRootType {
  countriesInformation: ICountryInformation[];
  exchangeRates: IExchangeRates[];
}

/** UTILS */
import mapFetchCountry from '@/utils/mapFetchCountry';

export const useCountryInformationStore = defineStore(
  'useCountryInformationStore',
  {
    state: (): IRootType => {
      return {
        countriesInformation: [],
        exchangeRates: [],
      };
    },
    actions: {
      async fetchPolishExchangeRates(): Promise<void> {
        try {
          const exchangeRatesA = await fetch(
            'https://api.nbp.pl/api/exchangerates/tables/A?format=json'
          );

          const exchangeRatesADate = await exchangeRatesA.json();
          const exchangeRatesB = await fetch(
            'https://api.nbp.pl/api/exchangerates/tables/B?format=json'
          );
          const exchangeRatesBDate = await exchangeRatesB.json();
          const exchangeRatesC = await fetch(
            'https://api.nbp.pl/api/exchangerates/tables/C?format=json'
          );
          const exchangeRatesCDate = await exchangeRatesC.json();

          if (
            exchangeRatesA.status !== 200 ||
            exchangeRatesB.status !== 200 ||
            exchangeRatesC.status !== 200
          ) {
            throw new Error("Can't fetch exchange rates information");
          }
          this.exchangeRates = [
            ...exchangeRatesADate[0].rates,
            ...exchangeRatesBDate[0].rates,
            ...exchangeRatesCDate[0].rates,
          ];
        } catch (error) {
          this.exchangeRates = [];
          console.error('fetchPolishExchangeRates', error);
        }
      },
      async fetchCountryById(id: string, polishCountryName: string) {
        try {
          const foundedCountry = this.countriesInformation.find(
            (country: ICountryInformation) =>
              country.id?.value.toLowerCase() === id.toLowerCase()
          );

          if (foundedCountry) {
            return foundedCountry;
          }
          const fetchData = await fetch(
            `https://restcountries.com/v3.1/alpha/${id}`
          );
          if (fetchData.status !== 200) {
            throw new Error("Can't fetch country information");
          }
          const countryDate = await fetchData.json();

          const countryInformation: ICountryInformation = mapFetchCountry(
            countryDate,
            polishCountryName
          );

          this.countriesInformation.push(countryInformation);

          return countryInformation;
        } catch (error) {
          console.error('fetchCountryById', error);
          return undefined;
        }
      },
    },
  }
);
