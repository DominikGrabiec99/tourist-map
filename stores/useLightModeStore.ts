import { defineStore } from 'pinia';

import EPageLightMode from '@/ts/enums/PageLightMode';
import EStorage from '@/ts/enums/Storage';

interface IRootType {
  lightMode: EPageLightMode | string;
}

export const useLightModeStore = defineStore('lightMode', {
  state: (): IRootType => {
    return {
      lightMode: '',
    };
  },
  getters: {
    getLightMode: (state): EPageLightMode | string => {
      return state.lightMode || EPageLightMode.LIGHT;
    },
  },
  actions: {
    toggleLightMode(): void {
      if (process.server) return;

      if (this.lightMode === EPageLightMode.LIGHT) {
        this.lightMode = EPageLightMode.DARK;
      } else {
        this.lightMode = EPageLightMode.LIGHT;
      }
      this.setLightModeOnPage();
    },
    getLightModeFromStorage() {
      if (process.server) return;
      this.lightMode =
        localStorage.getItem(EStorage.THEME) || EPageLightMode.LIGHT;
    },
    setLightModeOnPage(): void {
      if (process.server) return;
      localStorage.setItem(EStorage.THEME, this.lightMode || '');
      if (this.lightMode === EPageLightMode.DARK) {
        document.documentElement.classList.add('dark');
      } else {
        document.documentElement.classList.remove('dark');
      }
    },
  },
});
