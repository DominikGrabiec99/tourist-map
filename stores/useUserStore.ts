import { defineStore } from "pinia";

interface IState {
  user: Object | undefined
}

export const useUserStore = defineStore('userStore', {
  state: (): IState => {
    return {
      user: undefined
    }
  },
  getters: {
    getUser(state) {
      return state.user
    }
  }
})