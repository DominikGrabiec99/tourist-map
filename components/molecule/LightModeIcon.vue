<template>
  <Icon
    :name="mode[getLightMode]"
    size="24"
    class="dark:text-white pl-1 cursor-pointer transition-color"
    @click="toggleLightMode"
  />
</template>

<script setup lang="ts">
import { useLightModeStore } from '@/stores/useLightModeStore';
import { storeToRefs } from 'pinia';

import EPageLightMode from '@/ts/enums/PageLightMode';

const lightModeStore = useLightModeStore();

const { toggleLightMode } = lightModeStore;

const mode = {
  [EPageLightMode.DARK]: 'solar:sun-2-broken',
  [EPageLightMode.LIGHT]: 'solar:moon-broken',
}

const { getLightMode } = storeToRefs(lightModeStore);
</script>
