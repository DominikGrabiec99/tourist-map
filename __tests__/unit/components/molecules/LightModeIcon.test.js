import { describe, it } from 'vitest';
import { setActivePinia, createPinia } from 'pinia';
import { mount } from '@vue/test-utils';

import { useLightModeStore } from '@/stores/useLightModeStore';

import EPageLightMode from '@/ts/enums/PageLightMode';
import EStorage from '@/ts/enums/Storage';

import LightModeIcon from '@/components/molecule/LightModeIcon.vue';

describe('LightModeIcon.vue', () => {
  let wrapper;
  let lightModeStore;

  beforeAll(() => {
    setActivePinia(createPinia());
    wrapper = mount(LightModeIcon);
  });

  beforeEach(() => {
    lightModeStore = useLightModeStore();
  });

  afterEach(() => {
    lightModeStore.$reset();
    localStorage.clear();
  });

  it('render component', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('check is light mode icon', () => {
    expect(wrapper.element.getAttribute('name')).toStrictEqual(
      'solar:moon-broken'
    );
  });

  it('check toggle mode', async () => {
    lightModeStore.getLightModeFromStorage();
    lightModeStore.setLightModeOnPage();
    await wrapper.trigger('click');
    expect(lightModeStore.lightMode).toStrictEqual(EPageLightMode.DARK);
    await wrapper.trigger('click');
    expect(lightModeStore.lightMode).toStrictEqual(EPageLightMode.LIGHT);
  });

  it('check is dark mode icon', async () => {
    localStorage.setItem(EStorage.THEME, EPageLightMode.DARK);
    lightModeStore.getLightModeFromStorage();
    lightModeStore.setLightModeOnPage();
    expect(wrapper.element.getAttribute('name')).toStrictEqual(
      'solar:moon-broken'
    );
  });
});
