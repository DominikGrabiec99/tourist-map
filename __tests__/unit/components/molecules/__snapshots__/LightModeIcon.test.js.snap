// Vitest Snapshot v1

exports[`LightModeIcon.vue > render component 1`] = `
<icon
  class="dark:text-white pl-1 cursor-pointer transition-color"
  name="solar:moon-broken"
  size="24"
/>
`;
